### PR TITLE
Fix the meta boxes toggle aria-expanded for Firefox+NVDA #2

### LINKED ISSUE #1556: Fix the sidebar meta boxes toggle aria-expanded reporting for Firefox+NVDA
See #1387 and #1388, then reverted in #1537.
Firefox and NVDA don't announce the `aria-expanded` state change (expanded/collapsed) on the sidebar meta boxes toggles. The previous attempt to fix this with HTML/CSS broke the icon buttons when used in other context, my bad.

However, it would be great to find an alternative workaround, since Firefox/NVDA users don't get absolutely any feedback when the `aria-expanded` state changes. Yes, this is a mix of browser and screen reader bug and, ideally, it should be fixed upstream. Unfortunately, the tickets related to the Firefox bug appear to be pending since a few years and I'm afraid a resolution won't happen so soon.

Fully understanding the issue is a good first step, some references:
https://github.com/nvaccess/nvda/issues/5247
https://github.com/nvaccess/nvda/issues/6401
https://bugzilla.mozilla.org/show_bug.cgi?id=659886
https://bugzilla.mozilla.org/show_bug.cgi?id=1285862

In a few words, when Firefox repaints some elements, for example because some CSS generated or SVG icon changes, it also updates the complete set of accessible properties (the "accessible") exposed to screen readers. It shouldn't do this.
Quoting from https://github.com/nvaccess/nvda/issues/6401
> Firefox is regenerating the accessible for that button when it expands/collapses. So, instead of firing a stateChange, it fires focus on a "new" accessible. See Mozilla bugs 659886 and 1285862.
> If it were a stateChange, NVDA would report it. Instead, NVDA suppresses reporting the focus event because it's at the same position in the document (and not doing this would result in a lot of duplication when moving around).

Based on this, I've experimented what seems to me a logical workaround: just make the two events happen at a different time. Slightly delaying the `aria-expanded` state update, seems to fix it because the "accessible" regeneration happen first, and then Firefox fires a stateChange.

PR incoming soon.

### PR DESCRIPTION
**Pull request by afercia**
Second attempt to try a workaround to make Firefox+NVDA correctly report the aria-expanded state change on the meta boxes toggles in the sidebar.
See #1387 and #1388, then reverted in #1537.

I know this is a workaround, but I'd say it's simple enough and it is an improvement for NVDA users.
It is also very specific to this toggle button context so, in a certain sense, it's "isolated".

- no HTML/CSS changes, so no visual breakage
- adds `type="button"` to the button, so no need for `event.preventDefault()`
- uses a separate state property for the `aria-expanded` attribute
- this property gets updated with a slight delay (more details on the issue)

Note: yes, 100ms is a totally arbitrary value. I've tried with smaller values but it didn't work.

Alternatively, as last resort, we could consider to don't use aria-expanded at all and announce the collapsed/expanded state with `wp.a11y.speak` instead.

Fixes #1556

### CODE REVIEW

**Comment by jasmussen** [comment]
Here's what happens to me on Windows, with Firefox and NVDA:

- I tab to a heading in the sidebar, when focus is on a heading, it reads aloud: "Heading, Open Section, Discussion button collapsed"
- I click space to open, and it reads aloud "Space, Open Section, Discussion button collapsed, Expanded".

If I click with the mouse on Discussion, it being collapsed from the start, it reads: "Heading, open section, discussion button collapsed, open section, discussion button collapsed, expanded".

In both cases it sounds a bit verbose to me, but I can't be completely sure if this is the intended behavior or not.

For what it's worth it seems better on Chrome, where clicking space to expand a section simply reads "Expanded".

**Comment by afercia** [comment]
Weird, I get a different behavior. Windows 10, Firefox 54, NVDA 2017.2 here.

screenshot 119

In Firefox + NVDA i don't get the heading content repeated. It just says "expanded" or "collapsed". I've also reset the NVDA configuration to the "factory defaults", just in case, and everything still worked fine.

Instead, with Firefox + JAWS 17 yes, I get the heading content announced again and it announces the whole content + previous status (e.g. "collapsed") and then the new status (e.g. "expanded").

Yes, ideally screen readers should announce just the state change. When, for any reason, they think a new control just got focus, they will report again the entire control.

Basically what we're trying to do here is to hack around a browser/screen reader bug. Not entirely convinced we should do this, even if the alternative (complete silence and no feedback at all) is definitely something to avoid 😞 . Will ask feedback to other a11y folks.

Suggestion:
when testing with NVDA I find very useful to keep the "Speech viewer" open to have the reported text visible in a floating window. Can be enabled form the NVDA icon in the Windows system tray:

screenshot 120

**Comment by rianrietveld** [comment]
@afercia I think this is better and more native ARIA solution and looks good to me.
Not so fond of throwing in `wp.a11y.speak()` if we don’t need to.

**Comment by afercia** [comment]
Just tested again with NVDA 2017.3 and both Firefox ESR and Firefox 57, seems this works correctly now.

**Comment by nandujee** [comment]
Hi Afercia,

Aria-expanded tested in the FF 59.0.2 Version + NVDA 2018.1.1. 
It is NOT working. 
Aria-expanded attribute added to the toggle button. On expand, the aria-expanded true value has been set but NVDA reads nothing.
Note: It is working fine in IE + JAWS

Please help me to overcome the issue.
Thank you,
Nanda

**Comment by afercia** [comment]
@nandujee I'm very sorry I haven't noticed your comment until now. We're still investigating on a proper solution. It happens for a mix of different reasons, partly Firefox and partly NVDA. There's an open issue on the NVDA repo, see https://github.com/nvaccess/nvda/issues/5247
